src/mercury-agent/mercury/agent/procedures/hp_firmware.py

```
@@ -0,0 +1,259 @@
+    """
+    hpsum_log_path = '/var/hp/log/localhost/hpsum_log.txt'
+    exit_code = None
+    with open(hpsum_log_path) as f:
+        contents = f.readlines()
+
+    for line in contents:
+        if 'Exit Code' in line:
+            exit_code = line.split(':')[1]
+
+    status = get_return_code_status(int(exit_code))
+    log.info(status.get('message'))
+    if status.get('error_state'):
+        raise HPFirmwareException('Flashing failed with exit code {0}'.format(
+            exit_code))
+    return status.get('reboot_flag')
+
+
+@capability('hp_apply_bios_settings',
+            description='Apply bios settings found in given file',
+            kwargs_name=['url'], serial=False,
+            dependency_callback=is_hp, timeout=60)
+def hp_apply_bios_settings(url=None):
+    """
+    Apply BIOS settings found at the given URL
+
+    :param url: Full URL to the BIOS file
+    """
+    # TODO: Implement this
+    # TODO: Alongside the conrepfile, some blobs have to be installed to bring
+    # certain components up to a recommended version. Who determines which blobs
+    # to execute? Are they provided or does this method
+    # keep track of those based on the model 
+
+    # DL380Gen9 = HPServer(name="ProLiant DL380 Gen9",
+    #                      cfg='DL380Gen9_conrep_RAX.dat',
+    #                      blobs=['BCM5719_2.13.5_RHEL6.scexe',
+    #                             'HPSA_1.18_RHEL6.scexe',
+    #                             'iLO4_2.02_RHEL6.scexe'])
+
+
+@capability('hp_update_firmware',
+            description='Installs updates from provided packages',
+            kwarg_names=['url', 'dry_run'], serial=False,
+            dependency_callback=is_hp, timeout=3600)
+def hp_update_firmware(url=None, dry_run=False):
+    """
+    Apply HP firmware updates
+
+    :param url: Full URL to the package containing firmware files
+    :param dry_run: If firmware updates should be applied when found
+    """
+    if not url:
+        return
```

On this note, we should probably think about adding type safety to the RPC. It's not there now, well, because python programmers are anarchists. 

For now, would recommend not logging.. and not checking.. just let the thing blow up and bubble an exception if someone didn't provide url